// @salesforce/apex-node/lib/src/narrowing/narrowing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidApexClassID = exports.isValidTestRunID = exports.isNotQuotable = exports.isEmpty = exports.isTableConfig = exports.isTitle = exports.isColumn = exports.isRow = exports.isPrimitive = exports.isNull = exports.isArray = exports.isObject = exports.isNumber = exports.isBoolean = exports.isString = exports.isTestResult = void 0;
const constants_1 = require("../tests/constants");
const tsTypes = require("@salesforce/ts-types");
const isTestResult = (result) => {
    return (result.summary !== undefined &&
        result.tests !== undefined);
};
exports.isTestResult = isTestResult;
exports.isString = tsTypes.isString, exports.isBoolean = tsTypes.isBoolean;
const isNumber = (value) => {
    return typeof value === 'number';
};
exports.isNumber = isNumber;
const isObject = (value) => {
    return value !== null && typeof value === 'object' && !Array.isArray(value);
};
exports.isObject = isObject;
const isArray = (value) => {
    return Array.isArray(value);
};
exports.isArray = isArray;
const isNull = (value) => {
    return value === null;
};
exports.isNull = isNull;
const isPrimitive = (value) => {
    return ((0, exports.isString)(value) || (0, exports.isNumber)(value) || (0, exports.isBoolean)(value) || (0, exports.isNull)(value));
};
exports.isPrimitive = isPrimitive;
const isRow = (obj) => {
    if (typeof obj !== 'object' || obj === null) {
        return false;
    }
    for (const key in obj) {
        if (typeof key !== 'string' || typeof Reflect.get(obj, key) === 'string') {
            return false;
        }
    }
    return true;
};
exports.isRow = isRow;
const isColumn = (obj) => {
    return (typeof obj === 'object' &&
        obj !== null &&
        typeof Reflect.get(obj, 'key') === 'string' &&
        typeof Reflect.get(obj, 'label') === 'string');
};
exports.isColumn = isColumn;
const isTitle = (obj) => {
    return (typeof obj === 'object' &&
        obj !== null &&
        typeof Reflect.get(obj, 'title') === 'string');
};
exports.isTitle = isTitle;
const isTableConfig = (obj) => {
    return ((0, exports.isObject)(obj) &&
        Reflect.has(obj, 'title') &&
        (0, exports.isArray)(Reflect.get(obj, 'columns')) &&
        Reflect.get(obj, 'columns').length > 0);
};
exports.isTableConfig = isTableConfig;
const isEmpty = (value) => {
    return (value === null ||
        value === undefined ||
        (typeof value === 'string' && value.length === 0));
};
exports.isEmpty = isEmpty;
const isNotQuotable = (value) => (0, exports.isNull)(value) || (0, exports.isBoolean)(value) || (0, exports.isNumber)(value);
exports.isNotQuotable = isNotQuotable;
function isValidTestRunID(testRunId) {
    return ((testRunId.length === 15 || testRunId.length === 18) &&
        testRunId.startsWith(constants_1.TEST_RUN_ID_PREFIX));
}
exports.isValidTestRunID = isValidTestRunID;
function isValidApexClassID(apexClassId) {
    return ((apexClassId.length === 15 || apexClassId.length === 18) &&
        apexClassId.startsWith(constants_1.CLASS_ID_PREFIX));
}
exports.isValidApexClassID = isValidApexClassID;
