// @oclif/plugin-command-snapshot/lib/snapshot-command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapshotCommand = void 0;
const core_1 = require("@oclif/core");
const _ = require("lodash");
class SnapshotCommand extends core_1.Command {
    get commands() {
        const devPlugins = this.config.pjson.oclif.devPlugins ?? [];
        const commands = this.config.commands
            // Ignore dev plugins
            .filter(command => !devPlugins.includes(command.pluginName ?? ''))
            // remove aliases that reference themselves
            .filter(command => !command.aliases.includes(command.id));
        return _.sortBy(commands, 'id');
    }
    get entries() {
        return this.commands.map(command => {
            return {
                command: command.id,
                plugin: command.pluginName,
                flags: Object.keys(command.flags).sort(),
                alias: command.aliases,
                flagChars: Object.values(command.flags).map(flag => flag.char).filter(char => char).sort(),
                flagAliases: Object.values(command.flags).flatMap(flag => flag.aliases).filter(alias => alias).sort(),
            };
        });
    }
}
exports.SnapshotCommand = SnapshotCommand;
